src/common.py

```diff
@@ -1,33 +1,142 @@
+        # try except added to prevent falling into error when scalars_0 has 0 fields
+        try:
+            temp_list = [header_type["fields"][0]]
+            for i in range(1, len(header_type["fields"])):
+                if (temp_list[-1][0][:4] == "_pad"):
+                    temp_list = temp_list[:-1]
+                    temp_list.append(
+                        [header_type["fields"][i][0], header_type["fields"][i-1][1]+header_type["fields"][i][1]])
                 else:
                     temp_list.append(header_type["fields"][i])
             header_type["fields"] = temp_list
         except:
             pass
-        
+
     return data
 
-# assign valid name to state depending on which header it extracts
+
+def nibble(size):
+    if (size <= 8):
+        return 2
+    if (size <= 16):
+        return 4
+    if (size <= 24):
+        return 6
+    if (size <= 32):
+        return 8
+    if (size <= 40):
+        return 10
+    if (size <= 48):
+        return 12
+    if (size <= 56):
+        return 14
+    if (size <= 64):
+        return 16
+    return "-- fill blank here"
+
+
+def delete_obj(del_list, orig_list):
+    for item in del_list:
+        orig_list.remove(item)
+
+
+class State:
+    def __init__(self, name):
+        self.name = name
+        self.children = []
+
+    def print_state(self):
+        print("node's name: ", self.name)
+        print("node's children: ", [
+              self.children[i].name for i in range(len(self.children))])
+
+
+def find_children(root, nodes):
+    if len(nodes) == 0:
+        return
+    else:
+        children = []
+        del_list = []
+        for node in nodes:
+            if node[0] == root.name:
+                children.append(node[-1])
+                del_list.append(node)
+        delete_obj(del_list, nodes)
+        children = set(children)
+        for child in children:
+            state = State(child)
+            find_children(state, nodes)
+            root.children.append(state)
+        return
+
+
+def make_tree(graph):
+    paths = []
+    state_names = [edge[0] for edge in graph]
+    state_names = set(state_names)
+    non_roots = [edge[-1] for edge in graph]
+    non_roots = set(non_roots)
+    for name in state_names:
+        if name not in non_roots:
+            root = State(name)
+            find_children(root, graph)
+            paths.append(root)
+            root.print_state()
+    return paths
+
+
+def find_eth_subhdr(node, sub_headers):
+    if len(node.children) == 0:
+        if node.name != "final":
+            sub_headers.append(node.name)
+        return
+    else:
+        for child in node.children:
+            if child.name != "final":
+                sub_headers.append(child.name)
+            find_eth_subhdr(child, sub_headers)
+        return
+
+
+def find_ethernet(node, rmv_headers, sub_headers, ehter_detect):
+    if (node.name == "ethernet" or node.name == "Ether") and ehter_detect:
```

Comment:
This condition was inconsistent in two out of three code paths from where this has been modularized